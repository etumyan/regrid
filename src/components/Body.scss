@import '../styles/variables';

.regrid-body {
  overflow: auto;
  background-color: #fbfbfb;
  border-width: 1px;
  border-style: solid;
  border-color: $line-color;
  z-index: 10;

  &__inner {
    display: inline-flex;
    flex-direction: column;
    min-height: 100%;
    background-color: $line-color;
    box-shadow: 0 1px 0 $line-color,
                0 -1px 0 $line-color;
  }
}
