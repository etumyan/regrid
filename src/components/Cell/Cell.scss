@import '../../styles/variables';

.regrid-cell {
  flex-grow: 1;
  flex-basis: 0;
  margin-right: 1px;
  padding: 8px;
  min-width: 0;
  width: auto;
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
  cursor: default;

  &:last-child {
    margin-right: 0;
  }

  &_fixed-width {
    flex: 0 0 auto; // #TODO hack for defaults
  }

  &_align_start {
    justify-content: flex-start;
    text-align: left;
  }

  &_align_center {
    justify-content: center;
    text-align: center;
  }

  &_align_end {
    justify-content: flex-end;
    text-align: right;
  }
}
