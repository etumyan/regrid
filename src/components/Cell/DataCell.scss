@import '../../styles/variables';
@import './Cell.scss';

.regrid-cell_type_data {
  background-color: #fff;

  a {
    color: #208cd0;
    text-decoration: none;
  }
}
