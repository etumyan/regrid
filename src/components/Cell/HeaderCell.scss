@import '../../styles/variables';
@import './Cell.scss';

.regrid-cell_type_header {
  position: relative;
  background-color: #efefef;
  box-shadow: inset 1px 1px 0 #fff;
  user-select: none;

  &.regrid-cell_sortable {
    padding-right: 24px;
    position: relative;
    cursor: pointer;

    &::before,
    &::after {
      display: block;
      position: absolute;
      top: 50%;
      right: 8px;
      content: '';
      border-width: 4px;
      border-style: solid;
    }

    &::before {
      margin-top: -8px;
      border-color: transparent transparent #dcdcdc;
    }

    &::after {
      margin-top: 2px;
      border-color: #dcdcdc transparent transparent;
    }

    &.regrid-cell_align_center {
      text-indent: 16px;
    }
  }

  &.regrid-cell_sort_asc::before {
    border-color: transparent transparent #9e9e9e;
  }

  &.regrid-cell_sort_desc::after {
    border-color: #9e9e9e transparent transparent;
  }
}
