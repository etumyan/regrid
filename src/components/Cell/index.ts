export * from './Cell';
export * from './HeaderCell';
export * from './DataCell';
