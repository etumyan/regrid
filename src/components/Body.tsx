import './Body.scss';

import React, { FunctionComponent, useState, useEffect } from 'react';

import { hasElementHorizontalScrollbar } from '../utils';
import { Column, Sorting, DataSource, DataSourceItem } from './DataGrid';
import { DataRow } from './Row';

interface BodyProps {
  dataSource: DataSource;
  sorting?: Sorting;
  pageSize: number;
  currentPageIndex: number;
  primaryKey: Column['key'];
  columns: Column[];
  selectedRows: Set<Column['key']>;
  onRowClick: (key: Column['key']) => void;
}

export const Body: FunctionComponent<BodyProps> = ({
  dataSource,
  sorting,
  pageSize,
  currentPageIndex,
  primaryKey,
  columns,
  selectedRows,
  onRowClick
}) => {
  const [hasScrollbar, setHasScrollbar] = useState(false);

  useEffect(() => {
    setHasScrollbar(hasElementHorizontalScrollbar('.regrid-body'));
  });

  const offset = pageSize * currentPageIndex;
  const rows = [...dataSource].sort((a, b) => sort(a, b, sorting)).slice(
    offset, offset + pageSize
  ).map(row => (
    <DataRow
      key={row[primaryKey]}
      cells={columns}
      data={row}
      selected={selectedRows.has(row[primaryKey])}
      onClick={onRowClick.bind(
        undefined,
        primaryKey ? row[primaryKey] : undefined
      )}
    />
  ));

  return (
    <div className="regrid-body">
      <div className="regrid-body__inner">
        {rows.length ? rows : <div className="no-data">No data</div>}
        <DataRow
          cells={columns.map(column => ({ ...column, render: undefined }))}
          data={[]}
        />
      </div>
    </div>
  );
};

function sort(a: DataSourceItem, b: DataSourceItem, sorting: any) {
  let result = 0;

  if (!sorting) return result;

  const sortBy = sorting[0];
  const sortOrder = sorting[1].toLowerCase();

  if (a[sortBy] > b[sortBy]) {
    result = sortOrder === 'desc' ? -1 : 1;
  }

  if (a[sortBy] < b[sortBy]) {
    result = sortOrder === 'desc' ? 1 : -1;
  }

  return result;
}
