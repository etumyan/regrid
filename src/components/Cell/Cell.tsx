import './Cell.scss';

import React, { FunctionComponent } from 'react';

export interface CellProps {
  id: string | number;
  content?: string;
  width?: string | number;
  align?: 'start' | 'center' | 'end';
}

export const Cell: FunctionComponent<CellProps> = () => {
  return <React.Fragment />;
};
