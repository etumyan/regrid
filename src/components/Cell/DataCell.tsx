import './DataCell.scss';

import React, { FunctionComponent } from 'react';
import classNames from 'classnames';

import { CellProps } from './Cell';

interface DataCellProps extends CellProps {
  render?: (value: any, rowData: any) => string;
}

export const DataCell: FunctionComponent<DataCellProps> = ({
  id, content, width, align, render
}) => {
  const classNameList = [
    'regrid-cell',
    'regrid-cell_type_data',
    `regrid-cell_key_${id}`
  ];

  if (width !== undefined && width !== 'auto') {
    classNameList.push('regrid-cell_fixed-width');
  }

  switch (align) {
    case 'start':
      break;

    case 'center':
      classNameList.push('regrid-cell_align_center');
      break;

    case 'end':
      classNameList.push('regrid-cell_align_end');
      break;
  }

  const htmlContentWrapper = {
    __html: (render ? render(content, {}) : content) || ''
  };

  return (
    <div
      className={classNames(classNameList)}
      style={width ? { width } : undefined}
      dangerouslySetInnerHTML={htmlContentWrapper}
    >
    </div>
  );
};
