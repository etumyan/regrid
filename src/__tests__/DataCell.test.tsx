import React from 'react';
import renderer from 'react-test-renderer';

import { DataCell } from '../components/Cell';

test('DataCell renders correctly', () => {
  const component = renderer.create(
    <DataCell id="id"></DataCell>,
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
