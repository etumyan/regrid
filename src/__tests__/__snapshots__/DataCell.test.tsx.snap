// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataCell renders correctly 1`] = `
<div
  className="regrid-cell regrid-cell_type_data regrid-cell_key_id"
  dangerouslySetInnerHTML={
    Object {
      "__html": "",
    }
  }
/>
`;
